### PR TITLE
Folkengine/k01 validate charging station max profile

### PR DESCRIPTION
**Pull request by folkengine**
## Describe your changes

Handles K01.FR.38 from the OCPP 2.0.1 specification. 

> When chargingProfilePurpose = ChargingStationMaxProfilechargingProfileKind SHALL NOT be Relative

## Issue ticket number and link

[K01 - Charging Station Max Profile Validation #99](https://github.com/US-JOET/base-camp/issues/99)

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

**Comment by folkengine** [comment]
@shankari please review.

**Comment by shankari** [comment]
@folkengine this has a merge conflict

**Comment by folkengine** [comment]
> @folkengine this has a merge conflict

@shankari fixed.

File: lib/ocpp/v201/smart_charging.cpp [hunk 1]
```diff
@@ -21,6 +21,19 @@ ProfileValidationResultEnum SmartChargingHandler::validate_evse_exists(int32_t e
                                               : ProfileValidationResultEnum::Valid;
 }
 
+ProfileValidationResultEnum
+SmartChargingHandler::validate_charge_point_max_profile(const ChargingProfile& profile) const {
```
Comment (shankari) [inline lib/ocpp/v201/smart_charging.cpp:25]: @folkengine I am confused. I see the new method, but where is it invoked from?

File: lib/ocpp/v201/smart_charging.cpp [hunk 2]
```diff
@@ -21,6 +21,19 @@ ProfileValidationResultEnum SmartChargingHandler::validate_evse_exists(int32_t e
...
+    if (profile.chargingProfilePurpose != ChargingProfilePurposeEnum::ChargingStationMaxProfile) {
+        return ProfileValidationResultEnum::InvalidProfileType;
+    }
```
Comment (shankari) [inline lib/ocpp/v201/smart_charging.cpp:28]: Per the spec, `ChargingStationMaxProfile` can only be set on EVSE ID 0 (section 3.2, page 238). Why are we not checking for that in the validation? And why do we assume that the only purpose can be `MaxProfile`?

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp [hunk 3]
```diff
@@ -320,6 +340,56 @@ TEST_F(ChargepointTestFixtureV201, K01FR35_IfChargingSchedulePeriodsAreNotInChon
     EXPECT_THAT(sut, testing::Eq(ProfileValidationResultEnum::ChargingSchedulePeriodsOutOfOrder));
 }
 
+TEST_F(ChargepointTestFixtureV201, ValidateChargingStationMaxProfile_NotChargingStationMaxProfile_Invalid) {
```
Comment (shankari) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:343]: You need to fix the test name here to match the convention

**Review by shankari — CHANGES_REQUESTED** [review]

File: lib/ocpp/v201/smart_charging.cpp (hunk 1 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp:25]: Once the individual validations are written they will be added to a switchboard style general function that includes validations that apply to all profiles. For this card, it doesn't matter. That's TBD.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp (hunk 3 above)
Comment (folkengine) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:343]: Updated. This specific test is a general validation that doesn't have a specific FR.XX value.

File: lib/ocpp/v201/smart_charging.cpp (hunk 2 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp:28]: This is a really good callout. I was only looking at the specific reqs from K01. Updated the method signature to match what's being done for `validate_tx_profile()` and added the test `K04FR04_ValidateChargingStationMaxProfile_EvseIDgt0_Invalid`.

**Review by shankari — APPROVED** [review]

**Pull request closed without merging**